### PR TITLE
Update to JupyterLab 3.0

### LINKED ISSUE #100: Update for JupyterLab 3.0
Right now the documentation for JupyterLab points to a Binder link that installs 2.X, not 3.X!

### PR DESCRIPTION
**Pull request by jtpio**
Replaces https://github.com/jupyterlab/jupyterlab-demo/pull/101

Fixes #100 

For now commenting out some extensions that have not been updated yet.

Can be tested with this link: https://mybinder.org/v2/gh/jupyterlab/jupyterlab-demo/lab3?urlpath=lab

### TODO

- [x] Update to JupyterLab 3.0
- [x] `ipyleaflet`
- [x] `bqplot`
- [x] `jupyterlab-renderers`
- [ ] Check the predefined workspace is correctly restored

### CODE REVIEW

File: binder/environment.yml [hunk 1]
```diff
@@ -2,34 +2,34 @@ channels:
 - conda-forge
 dependencies:
 # Jupyter
-- jupyterlab=2
-- nodejs=11
-- nbconvert=5.5
-- widgetsnbextension=3.5
```
Comment (jasongrout) [inline binder/environment.yml:8]: If we keep this line, I think we'll still be able to open a widgets notebook in classic notebook from the jlab menu.

Comment (jasongrout) [reply binder/environment.yml:8]: Oh wait, never mind - it should be pulled in by ipywidgets.

File: binder/environment.yml [hunk 2]
```diff
@@ -2,34 +2,34 @@ channels:
...
-- dask=2.1
-- matplotlib=3.1
-- pandas=0.24
+- xeus-python=0.9
+- ipywidgets=7.6
+# - ipyleaflet=0.12
+# - altair=3.1
+# - bqplot=0.12
+# - dask=2.1
+# - matplotlib=3.1
```
Comment (jasongrout) [inline binder/environment.yml:16]: Don't we want matplotlib? It's ipympl that hasn't been updated.

Comment (jtpio) [reply binder/environment.yml:16]: Sure let add it back.

**Comment by jtpio** [comment]
Node 14 is giving the following:

image

Trying with Node 12 instead.

**Comment by jasongrout** [comment]
I'm updating the docs to use https://mybinder.org/v2/gh/jupyterlab/jupyterlab-demo/8f2ecb6?urlpath=lab/tree/demo in https://github.com/jupyterlab/jupyterlab/pull/9546

**Comment by jasongrout** [comment]
After a few seconds on that image, I get this dialog:

Build Recommended
JupyterLab build is suggested:

jupyter-offlinenotebook needs to be included in build

Can we remove that plugin too, or rebuild as part of the image?

**Comment by jasongrout** [comment]
> Can we remove that plugin too, or rebuild as part of the image?

I can't see where we are pulling in that extension. Is it part of the base binder image?

**Comment by manics** [comment]
> I can't see where we are pulling in that extension. Is it part of the base binder image?

Yes, it's in the default repo2docker scripts. I'm going to update it to a new version that works with both JL 2 and 3, and use the new pre-built extensions system so it won't require a rebuild for JL3.

**Comment by jtpio** [comment]
Yes it is part of the base image and installed here:

https://github.com/jupyterhub/repo2docker/blob/560b1d96a0e39cb8de53cb41a7c2d8d23384eb82/repo2docker/buildpacks/conda/install-miniforge.bash#L60

When the switch to JupyterLab 3.0 is done in repo2docker (https://github.com/jupyterhub/repo2docker/pull/996) this should normally not be an issue anymore since the offline notebook extension will be prebuilt.

**Comment by jtpio** [comment]
Wondering why the popup still shows up though, since there should also be an explicit build right after:

https://github.com/jupyterhub/repo2docker/blob/560b1d96a0e39cb8de53cb41a7c2d8d23384eb82/repo2docker/buildpacks/conda/install-miniforge.bash#L61

**Comment by jasongrout** [comment]
Probably because it is never incorporated into lab 3 or something?

We could install the newest version of that package, which supports jlab 3 now.

**Comment by jtpio** [comment]
The latest commit explicitly installs `jupyter-offlinenotebook` for now to avoid having the popup:

https://mybinder.org/v2/gh/jupyterlab/jupyterlab-demo/3818244?urlpath=lab

**Comment by jasongrout** [comment]
> https://mybinder.org/v2/gh/jupyterlab/jupyterlab-demo/3818244?urlpath=lab

I updated https://github.com/jupyterlab/jupyterlab/pull/9546 to use this new link

File: binder/environment.yml [hunk 3]
```diff
@@ -2,34 +2,35 @@ channels:
...
 - invoke=1.2
 # C++ Kernel
-- xeus-cling=0.6
+- xeus-cling=0.10
 - xtensor=0.20.8
 - xtensor-blas=0.16.1
-- xwidgets=0.18.0
-- xleaflet=0.9.0
+- xwidgets=0.24.2
+- xleaflet=0.13.1
```
Comment (SylvainCorlay) [inline binder/environment.yml:33]: I think we can now include `ipyleaflet=0.13.6` which has jupyterlab 3.0 support, and `xeus-cling=0.10.1`.

Comment (jtpio) [reply binder/environment.yml:33]: Nice!

**Comment by jtpio** [comment]
Looks good with the jupyterlab renderers:

image

I think it's mostly `bqplot` left now. Linking to the PR for reference: https://github.com/bqplot/bqplot/pull/1222

**Comment by jtpio** [comment]
Also it seems like the Binder doesn't open with the predefined workspace anymore.

So something to have a look at too, also to make sure it's not a regression.

**Comment by jtpio** [comment]
For anyone willing to test the latest changes, here is the link to the Binder for this PR:

https://mybinder.org/v2/gh/jupyterlab/jupyterlab-demo/lab3?urlpath=lab

File: binder/environment.yml [hunk 4]
```diff
@@ -2,34 +2,36 @@ channels:
...
-- seaborn=0.9
+- seaborn=0.11
 - tensorflow=1.13
 - sympy=1.4
 - pyyaml
 - traittypes==0.2.1
 - invoke=1.2
 # C++ Kernel
-- xeus-cling=0.6
+- xeus-cling=0.10
```
Comment (jtpio) [inline binder/environment.yml:30]: To answer your question @SylvainCorlay, this syntax installs `0.10.1`:

image

**Comment by jtpio** [comment]
> Also it seems like the Binder doesn't open with the predefined workspace anymore.

This issue might be relevant: https://github.com/jupyterlab/jupyterlab/issues/9186

**Comment by jtpio** [comment]
Just added `bqplot` back, this is starting to look good.

image


Apart from the workspace restoration issue mentioned above, everything should work as before.

We might also want to remove `jupyter-offlinenotebook=0.2` from the repo once it is part of `repo2docker` by default: https://github.com/jupyterhub/repo2docker/pull/997 (this could be a separate PR)

**Comment by jtpio** [comment]
> Apart from the workspace restoration issue mentioned above, everything should work as before.

Just opened https://github.com/jupyterlab/jupyterlab/issues/9590 as it looks like there might be an issue with the restorer?

**Comment by jtpio** [comment]
Marking as ready for review, so the workspace issue could be addressed separately.

Instead, we can update the URL to open the `Lorenz.ipynb` notebook by default using the following link:

https://mybinder.org/v2/gh/jupyterlab/jupyterlab-demo/lab3?urlpath=lab/tree/demo/Lorenz.ipynb

**Review by saulshanabrook — APPROVED** [review]
This looks good to me!

I tried out this branch and seems good.

I agree with @jtpio, to merge this and not block on the upstream JupyterLab 3.0 workspace issue (https://github.com/jupyterlab/jupyterlab/issues/9590).

**Comment by blink1073** [comment]
I agree, thanks @saulshanabrook!

**Pull request merged**